types/mempool/mempool.go

```diff
@@ -0,0 +1,409 @@
+package mempool
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/cosmos/cosmos-sdk/types"
+	"math"
+
+	huandu "github.com/huandu/skiplist"
+
+	"github.com/cosmos/cosmos-sdk/x/auth/signing"
+)
+
+// Tx we define an app-side mempool transaction interface that is as
+// minimal as possible, only requiring applications to define the size of the
+// transaction to be used when reaping and getting the transaction itself.
+// Interface type casting can be used in the actual app-side mempool implementation.
+type Tx interface {
+	types.Tx
+
+	// Size returns the size of the transaction in bytes.
+	Size() int
+}
+
+// HashableTx defines an interface for a transaction that can be hashed.
+type HashableTx interface {
```

Comment:
nit: no need for a separate interface -- just update the base `Tx` interface and be sure to update the ADR :)